Add query args to be able to modify the current DB

andresmgot opened the pull request:
Fixes #20

This PR allows to change the datasource database based on the `Query` `Args`. At the moment, this is implemented as a backward-compatible change. If the datasource implements the `DB` function, that will be used to retrieve the database, passing the query as an argument. If not the behavior is the same than before.

Note that this is something generic, it make sense, for a datasource, to send the database name as part of the query so people can reuse the same datasource credentials to connect to different databases. If we want to give this feature a more exposed support, we could change the [`Connect` interface](https://github.com/grafana/sqlds/blob/main/driver.go#L15) to add an optional parameter (the Query args) or add the `DB` method to the interface, but that would be a breaking change. WDYT?

kminehart commented on query.go:
```diff
@@ -25,8 +25,9 @@ const (
-	Format FormatQueryOption `json:"format"`
+	RawSQL string                 `json:"rawSql"`
+	Format FormatQueryOption      `json:"format"`
+	Args   map[string]interface{} `json:"args"`
```
Could we use `json.RawMessage` here instead? `map[string]interface{}` is so hard to work with.

kminehart reviewed (changes requested):
Just 1 request. Otherwise, LGTM! :)

This makes sense.  I'm curious to see how it could be used. My understanding is you typically what your database to be a sort of permanent fixture in your code, as they typically have some kind of connection pooling.

kminehart commented on datasource.go:
```diff
@@ -98,11 +99,20 @@ func (ds *sqldatasource) handleQuery(req backend.DataQuery) (data.Frames, error)
+	// The database connection may vary depending on query arguments
+	db := ds.db
+	if ds.DB != nil {
+		db, err = ds.DB(q)
```
We will probably need to close this database when the request is complete, right?

Just a simple `defer db.Close()` here might suffice.

andresmgot replied:
actually, it may be nice to reuse these db instances. Let me re-work this to support this (and do a clean-up in the `Dispose` method).

andresmgot commented:
> This makes sense. I'm curious to see how it could be used. My understanding is you typically what your database to be a sort of permanent fixture in your code, as they typically have some kind of connection pooling.

I tried this with some simple code:

```go
	ds := sqlds.NewDatasource(s)
	ds.DB = func(q *sqlds.Query) (*sql.DB, error) {
		return s.ConnectOverridingDefaults(q)
	}
```

We probably want to add some caching there so we don't create new connections for every request. That also means that we may need to handle resource disposal in a different way :thinking:

sunker reviewed (commented):
Nice way of solving this! 

I'm not sure the method name `DB` would make sense to devs that are new to this package. That being said, I don't have a better suggestion. Can we at least add some comments to the method? Also, I think lifting the DB method to the Connect interface and introduce a breaking change would be fine.

andresmgot commented:
> I'm not sure the method name `DB` would make sense to devs that are new to this package. That being said, I don't have a better suggestion. Can we at least add some comments to the method? Also, I think lifting the DB method to the Connect interface and introduce a breaking change would be fine.

At the end, I modified the `Connect` interface to include the optional args. It's a breaking change but it feels much more clean than the `DB` method (since both methods returned a `sql.DB`). Let me know what you think. I have also added a cache for the db connections and a proper disposal method.

kminehart commented on datasource.go:
```diff
@@ -82,6 +85,27 @@ func (ds *sqldatasource) QueryData(ctx context.Context, req *backend.QueryDataRe
 }
 
+func (ds *sqldatasource) getDB(q *Query) (*sql.DB, string, error) {
+	// The database connection may vary depending on query arguments
+	// The raw arguments are used as key to store the db connection in memory so they can be reused
+	key := ""
+	db := ds.dbConnections[key]
+	var err error
+	if len(q.Args) != 0 {
+		key = string(q.Args[:])
+		if cachedDB, ok := ds.dbConnections[key]; ok {
+			db = cachedDB
+		} else {
+			db, err = ds.c.Connect(ds.settings, q.Args)
+			if err != nil {
+				return nil, "", err
+			}
+			ds.dbConnections[key] = db
+		}
+	}
+	return db, key, nil
+}
```
```suggestion
	// The database connection may vary depending on query arguments
	// The raw arguments are used as key to store the db connection in memory so they can be reused
	key := DefaultKey
	db := ds.dbConnections[key]
	if len(q.Args) == 0 {
		return db, key, nil
	}
	
	key = string(q.Args[:])
	if cachedDB, ok := ds.dbConnections[key]; ok {
		return cachedDB, key, nil
	}
	
	db, err := ds.c.Connect(ds.settings, q.Args)
	if err != nil {
		return nil, "", err
	}
	
	// Assign this connection in the cache
	ds.dbConnections[key] = db
	
	return db, key, nil
}
```

Always prefer early returns over stacking else blocks.

kminehart reviewed (changes requested):
Here's some code related comments.

I'm struggling a little bit with the concept. I understand why it's needed but I'm a little worried about the potential for the amount of active connections to be really high.

It's really easy to reach the connection limit in Postgres, for example, which has a limit of around 100 concurrent connections. I don't know a lot about other SQL databases, but part of me worries that making it really easy to create one additional connection could be a little dangerous.

I think what'd make me happy with this change is if there was a way for things that satisfy the interface to intentionally avoid using the `dbConnections` map, and a comment explaining why you may want to do that.

andresmgot commented:
> I think what'd make me happy with this change is if there was a way for things that satisfy the interface to intentionally avoid using the `dbConnections` map, and a comment explaining why you may want to do that.
> 

That's a good point. Connections for AWS plugins are cached thanks to the aws-sdk package but it's not the general use. I will add a toggle for that.

kminehart commented on datasource.go:
```diff
@@ -88,6 +100,34 @@ func (ds *sqldatasource) QueryData(ctx context.Context, req *backend.QueryDataRe
+		return db.(*sql.DB), key, nil
+	}
+
+	var err error
+	key = string(q.ConnectionArgs[:])
+	if cachedDB, ok := ds.dbConnections.Load(key); ok {
+		return cachedDB.(*sql.DB), key, nil
+	}
+
+	db, err = ds.c.Connect(ds.settings, q.ConnectionArgs)
```
```suggestion
	key = string(q.ConnectionArgs[:])
	if cachedDB, ok := ds.dbConnections.Load(key); ok {
		return cachedDB.(*sql.DB), key, nil
	}

	var err error
	db, err = ds.c.Connect(ds.settings, q.ConnectionArgs)
```

kminehart replied:
nit but just moving `err` closer to where it's used

kminehart commented on datasource.go:
```diff
@@ -129,20 +175,25 @@ func (ds *sqldatasource) handleQuery(ctx context.Context, req backend.DataQuery)
 	if errors.Is(err, ErrorQuery) {
-		ds.db, err = ds.c.Connect(ds.settings)
+		db, err = ds.c.Connect(ds.settings, q.ConnectionArgs)
+		ds.dbConnections.Store(cacheKey, db)
```
should we be doing this before the error is checked on the next line?

kminehart commented on datasource.go:
```diff
@@ -88,6 +100,34 @@ func (ds *sqldatasource) QueryData(ctx context.Context, req *backend.QueryDataRe
+	}
+
+	var err error
+	key = string(q.ConnectionArgs[:])
```
```suggestion
	key = string(q.ConnectionArgs)
```

https://play.golang.org/p/uIM21uDBuJ_Y

kminehart reviewed (changes requested):
Just a couple minor changes!

Appreciate the PR.

kminehart reviewed (approved):
LGTM. Would you mind tagging this with a v2.0.0 tag? This PR and the most recent one both introduced breaking changes so we might as well bundle them in a major version change.

andresmgot commented:
> LGTM. Would you mind tagging this with a v2.0.0 tag? This PR and the most recent one both introduced breaking changes so we might as well bundle them in a major version change.

Thanks! Yes, will do that
